### PR TITLE
Basic mapper pass

### PR DESCRIPTION
**Pull request by 1ucian0**
### Summary

This pass is a mapper that implements a simplified swap mapper. By simplified, I mean, it does not try to keep the depth to the minimum. This means when two entangled wires are not directly connected in the coupling_map the shortest path is calculated, and the farthest wire is swap with the closest wire.

The pass parameters are:
  - `coupling_map` (`Coupling`): A coupling map to which the mapper will try to adapt the dag.
  - `swap_basis_element` (`string`): Default: `'swap'` the name of the gate that will be used for swaping.
  - `swap_data` (`dict`):  Default: `{"opaque": True, "n_args": 0, "n_bits": 2, "args": [], "bits": ["a", "b"]}`. The swap "gate data".

### Details and comments

I extended the `DAGCircuit` object with:
 - `get_cnot_nodes()`: Returns the nodes that are entangled.
 - `add_dag_at_the_end(dag, layout)`: adds `dag` at the end of `self`, using `layout` for wireing. If `layout` is not provided, creates a layout where wires are connected one-to-one.

I extened the `Coupling` object with:
 - `shortest_path`: returns the shorter path between two nodes in an undirected coupling graph.

### CODE REVIEW

File: qiskit/dagcircuit/_dagcircuit.py [hunk 1]
```diff
@@ -1341,6 +1352,30 @@ def property_summary(self):
                    "operations": self.count_ops()}
         return summary
 
+    def add_dag_at_the_end(self, dag, layout=None):
```
Comment (ajavadia) [inline qiskit/dagcircuit/_dagcircuit.py]: So how is this method different from `compose()`?

File: qiskit/dagcircuit/_dagcircuit.py [hunk 2]
```diff
@@ -1341,6 +1352,30 @@ def property_summary(self):
...
+        """Add `dag` at the end of `self`, using layout for wireing.
+        If `layout` is not provided, creates a layout where wires are
+        connected one-to-one."""
+        for qreg in dag.qregs.values():
+            with suppress(DAGCircuitError):
+                self.add_qreg(QuantumRegister(qreg.size, qreg.name))
```
Comment (ajavadia) [inline qiskit/dagcircuit/_dagcircuit.py]: Why do we need this suppress here? Can't you just do ```if qreg not in self.qregs```?

File: test/python/transpiler/test_swap_mapper.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        """No need to have any swap, the CX are distance 1 to each other
+         q0:--(+)-[U]-(+)-
+               |       |
+         q1:---.-------|--
+                       |
+         q2:-----------.--
+
+         Coupling map: [1]--[0]--[2]
+        """
+        coupling = Coupling({0: [1, 2]})
```
Comment (ajavadia) [inline test/python/transpiler/test_swap_mapper.py]: We are deprecating the "dict" style of coupling map, in favor of a "list" style. In #1141, @kdk has made a new Coupling constructor based on list. I think we should use that here too, and just stop using dict altogether for consistency.

File: qiskit/transpiler/passes/swap_mapper.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    """
+
+    def __init__(self, coupling_map, swap_basis_element='swap', swap_data=None):
+        """
+        Maps a DAGCircuit onto a `coupling_map` using swap gates.
+        Args:
+            coupling_map (Coupling): Directed graph represented a coupling map.
+            swap_basis_element (string):  Default: 'swap' the name of the gate
+               that will be used for swaping.
+            swap_data (dict): The swap "gate data". Default: the swap gate is opaque.
```
Comment (ajavadia) [inline qiskit/transpiler/passes/swap_mapper.py]: About the arguments to the SwapMapper's `__init__`:
- Why does it need a "swap_basis_element" and "swap_data"? Under what circumstances would I pass something other than the default here?

- Can we also make it take an `initial_layout` as well? We will soon call this a `desired_layout`. But basically, it's a just a layout that says if my program complies with this, don't change it.

File: qiskit/dagcircuit/_dagcircuit.py (hunk 1 above)
Comment (1ucian0) [reply qiskit/dagcircuit/_dagcircuit.py]: compose needs the two dags to have the same registers. This wraps compose to run `add_*reg` ahead.

File: qiskit/dagcircuit/_dagcircuit.py (hunk 2 above)
Comment (1ucian0) [reply qiskit/dagcircuit/_dagcircuit.py]: good point. Fixed in b27c2b1.

**Comment by 1ucian0** [comment]
> Why does it need a "swap_basis_element" and "swap_data"? Under what circumstances would I pass something other than the default here?

I think the pass should be agnostic of the specifics of the gate/arch. With this in mind, it seems to me reasonable that a pass that inserts gates X is also asking about the specifics about X. In this case, this pass inserts swaps, so give me info about swap. What do you think?

> Can we also make it take an initial_layout as well?

Sure.

> We will soon call this a desired_layout. But basically, it's a just a layout that says if my program complies with this, don't change it.

What about an analysis pass that checks if a layout complies? The swap pass can be conditional over that one.

File: qiskit/transpiler/passes/swap_mapper.py (hunk 4 above)
Comment (1ucian0) [reply qiskit/transpiler/passes/swap_mapper.py]: There are many talking points here. So, moving the the conversation to https://github.com/Qiskit/qiskit-terra/pull/1270#issuecomment-439951585

File: test/python/transpiler/test_swap_mapper.py (hunk 3 above)
Comment (1ucian0) [reply test/python/transpiler/test_swap_mapper.py]: The pass takes a `Coupling` object. It that class keeps the API, the pass should continue working. If the constructor changes, the test should be updated.

**Comment by ajavadia** [comment]
> I think the pass should be agnostic of the specifics of the gate/arch. With this in mind, it seems to me reasonable that a pass that inserts gates X is also asking about the specifics about X. In this case, this pass inserts swaps, so give me info about swap.

I think mappers on superconducting processors will always use swaps. That's how we bring two qubits next to each other. 

But even more generally, I think every component of Qiskit should be aware of the basic gate definitions within Qiskit, by just importing `qiskit.extensions.standard`. So I don't think we should worry about passing these basic definitions to different component, it will become very verbose.

> What about an analysis pass that checks if a layout complies? The swap pass can be conditional over that one.

I think this makes sense. We can reuse that pass for all other mappers. And the mappers don't have to worry about the `desired_layout` anymore. Just receive a circuit and a coupling map, and do their thing.

**Comment by 1ucian0** [comment]
Working when there is a single the qubit register.

File: test/python/transpiler/test_basic_mapper.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+         qr0:-------(+)---.--
+                     |    |
+         qr1:-----X--.---(+)-
+                  |
+         qr2:--X--|----------
+               |  |
+         qr3:--X--X----------
+
+        """
```
Comment (ajavadia) [inline test/python/transpiler/test_basic_mapper.py]: There seems to be a bug here.
This circuit is not the equivalent to the original. The second swap should be between `qr2` and `qr1` (that's how `qr3` moves up to `qr1` using a chain of swaps). I checked the branch and it does not produce that. Instead it produces this picture. 

Can you see if you can fix?

File: test/python/transpiler/test_basic_mapper.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        coupling = Coupling({0: [1, 2]})
+
+        qr = QuantumRegister(3, 'q')
+        circuit = QuantumCircuit(qr)
+        circuit.cx(qr[0], qr[1])
+        circuit.h(qr[0])
+        circuit.cx(qr[0], qr[2])
+
+        dag = circuit_to_dag(circuit)
+        before = deepcopy(dag)
```
Comment (ajavadia) [inline test/python/transpiler/test_basic_mapper.py]: no need for a deepcopy here. just assert that `after` and `dag` are the same. building `before` is redundant.

File: test/python/transpiler/test_basic_mapper.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        """
+        coupling = Coupling({0: [1], 1: [2], 2: [3]})
+
+        qr = QuantumRegister(4, 'q')
+        circuit = QuantumCircuit(qr)
+        circuit.cx(qr[2], qr[3])
+        circuit.cx(qr[0], qr[1])
+
+        dag = circuit_to_dag(circuit)
+        before = deepcopy(dag)
```
Comment (ajavadia) [inline test/python/transpiler/test_basic_mapper.py]: see above.

File: qiskit/transpiler/passes/mapping/basic_mapper.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+
+class BasicMapper(TransformationPass):
+    """
+    Maps (with minimum effort) a DAGCircuit onto a `coupling_map` adding swap gates.
+    """
+
+    def __init__(self,
+                 coupling_map,
+                 swap_gate=None,
```
Comment (ajavadia) [inline qiskit/transpiler/passes/mapping/basic_mapper.py]: Can you get rid of this argument? There is no reason to pass the `swap_gate`. A swap is always a swap.

File: qiskit/dagcircuit/_dagcircuit.py [hunk 9]
```diff
@@ -554,7 +554,20 @@ def _map_condition(self, wire_map, condition):
             new_condition = (wire_map.get(bit0, bit0)[0], condition[1])
         return new_condition
 
-    def compose_back(self, input_circuit, wire_map=None):
+    def extends_at_the_end(self, dag, edge_map=None):
```
Comment (ajavadia) [inline qiskit/dagcircuit/_dagcircuit.py]: this is highly similar to `compose_back()`. can we make the names similar too to indicate that?
```suggestion
    def extend_back(self, dag, edge_map=None):
```

File: test/python/transpiler/test_basic_mapper.py (hunk 5 above)
Comment (1ucian0) [reply test/python/transpiler/test_basic_mapper.py]: Correct. I fixed it in 49cb295c. Have a look.

File: qiskit/dagcircuit/_dagcircuit.py [hunk 10]
```diff
@@ -501,11 +501,11 @@ def _check_wiremap_registers(self, wire_map, keyregs, valregs, valreg=True):
                     # If mapping to a register not in valregs, add it.
                     # (k,0) exists in wire_map because wire_map doesn't
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/dagcircuit/_dagcircuit.py]: fixed in 162fb8b1

File: qiskit/dagcircuit/_dagcircuit.py [hunk 11]
```diff
@@ -456,7 +456,7 @@ def _make_union_gates(self, input_circuit):
                                       % k)
         return union_gates
 
-    def _check_wiremap_registers(self, wire_map, keyregs, valregs, valreg=True):
+    def _check_wiremap_registers(self, edge_map, keyregs, valregs, valreg=True):
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/dagcircuit/_dagcircuit.py]: fixed in 95816c59c96a3237c2d257d2b3ba1a271713ac95

File: test/python/transpiler/test_basic_mapper.py (hunk 6 above)
Comment (1ucian0) [reply test/python/transpiler/test_basic_mapper.py]: remove them in 66c33e06

**Comment by ajavadia** [comment]
Thanks for the rework. Overall it looks good, I'm just checking functionality with some random testing.
Can you check this? I think the H is misplaced here. After 2 and 3 are swapped, the H should be applied to 2.

```
from qiskit import QuantumRegister, ClassicalRegister, QuantumCircuit
from qiskit import Aer
from qiskit.transpiler import PassManager
from qiskit.transpiler.passes import BasicMapper, CXCancellation
from qiskit.transpiler import transpile
from qiskit.mapper import Coupling

coupling = {0: [1], 1: [2], 2: [3], 3: [4], 4: [5], 5: [6]}
qr = QuantumRegister(7, 'q')
circuit = QuantumCircuit(qr)
circuit.h(qr[3])
circuit.cx(qr[0], qr[6])
circuit.cx(qr[6], qr[0])

simulator = Aer.get_backend('qasm_simulator')
coupling_map = Coupling(couplingdict=coupling)
pass_manager = PassManager()
pass_manager.add_passes([BasicMapper(coupling_map=coupling_map)])
new_circ = transpile(circuit, simulator, pass_manager=pass_manager)
```

`circuit.draw(reverse_bits=True)`
image

`new_circ.draw(reverse_bits=True)`
image

**Comment by 1ucian0** [comment]
Good catch. Fixed it in 09c331dacbc8f4f2a59180697307d48de0bdc9e8.

**Pull request merged**